apache/cloudstack
# Fix data volume import / export functionality when using swift as secondary storage

### Description
This PR is a fix for the data volume import and export functionality when using swift as secondary storage.

Background:
Secondary storage is used for templates, snapshots and iso's. A data disk is a volume and is meant to reside on primary storage. However, when importing a volume it is first uploaded to secondary storage and then copied to over to primary storage when attached to a VM. Similarly when exporting a volume, it is copied to secondary storage before it is made available to the user for download. When using swift as a secondary storage, an intermediate staging nfs store is needed when copying volumes between primary and secondary storage. This staging nfs store is leveraged for the volume import and export functionality since volumes need not have a footprint on secondary storage.

Currently this functionality does not work:

When importing a volume the following error is observed:

`org.apache.cloudstack.storage.to.VolumeObjectTO cannot be cast to org.apache.cloudstack.storage.to.TemplateObjectTO
`

Further investigation reveals that there is no provision in the code to handle import of volumes. Currently only template import is supported.

When exporting a volume a URL to the secondary storage is generated but it clicking it does not initiate a download of the exported volume. Further investigation revealed that the url points to a broken symlink which points to an invalid path on the SSVM

Use Cases

1. User should be able to import a volume from a URL when using swift as secondary storage and XenServer as the hypervisor.
2. User should be able to export a volume to a secondary storage VM URL when using swift as secondary storage and XenServer as the hypervisor.















### Types of changes


### Feature/Enhancement Scale or Bug Severity
This addresses https://github.com/apache/cloudstack/issues/4874

#### Feature/Enhancement Scale
- [*] Minor

#### Bug Severity
- [*] Minor

### How Has This Been Tested?




1. Create data volume (non-managed storage)
2. Attach it to a VM.
3. Create a file system and a test file on the new data volume.
4. Export volume.
5. Upload exported volume to a bucket and generate a URL.
6. Import volume back into CloudStack from URL.
7. Attach imported disk to a VM.
8. Verify previously created test file.
